include/libtorrent/hash_picker.hpp

```diff
@@ -88,35 +88,38 @@ namespace libtorrent
 		std::int32_t node;
 	};
 
+	// the hash request represents a range of hashes in the merkle hash tree for
+	// a specific file ('file').
+	// base indicates which *level* of the tree we're referring to. 0 means the
+	// leaf level.
+	// index is the index of the first hash at the specified level.
+	// count is the number of hashes in the range
 	struct TORRENT_EXTRA_EXPORT hash_request
 	{
 		hash_request() = default;
 		hash_request(file_index_t const f, int const b, int const i, int const c, int const p)
 			: file(f), base(b), index(i), count(c), proof_layers(p)
-		{
-			TORRENT_ASSERT(file >= file_index_t{0} && base >= 0 && index >= 0 && count >= 0 && proof_layers >= 0);
-		}
+		{}
 
 		hash_request(hash_request const&) = default;
-		hash_request& operator=(hash_request const& o)
-		{
-			new (this) hash_request(o);
```

Comment:
this is fishy, and may actually be undefined behavior, since it changes values of variables declared `const`, so users of this object may assume that the members never change, even when they do here. Instead I made the members non-const. for immutable `hash_request` objects, I think we should just make those objects `const`

Reply:
Yeah. I've debated removing this myself. I really wanted to make `hash_request` always be immutable, but it turns out this is not easy to do in C++.